File: src/spectrace/registry/loader.py
"""YAML loader for the R0 document-local entity registry."""

from __future__ import annotations

from pathlib import Path

import yaml

from spectrace.registry.model import EntityRegistry, RegistryLoadError
from spectrace.registry.schema import parse_registry_data


def load_registry(path: str | Path) -> EntityRegistry:
    """Load a document-local registry from an explicit YAML path."""

    registry_path = Path(path)
    try:
        documents = tuple(
            yaml.safe_load_all(registry_path.read_text(encoding="utf-8"))
        )
    except yaml.YAMLError as error:
        raise RegistryLoadError(
            f"{registry_path} contains invalid YAML"
        ) from error

    if len(documents) != 1:
        raise RegistryLoadError(
            f"{registry_path} must contain exactly one YAML document"
        )

    [raw_data] = documents
    return parse_registry_data(raw_data)

File: src/spectrace/registry/loader.ts
import { readFile } from "node:fs/promises";

import { parseAllDocuments } from "yaml";

import { EntityRegistry, RegistryLoadError } from "./model.js";
import { parseRegistryData } from "./schema.js";

export async function loadRegistry(path: string): Promise<EntityRegistry> {
  let registryText: string;

  try {
    registryText = await readFile(path, "utf8");
  } catch (error) {
    throw new RegistryLoadError(`${path} cannot be read`, { cause: error });
  }

  let documents: ReturnType<typeof parseAllDocuments>;

  try {
    documents = parseAllDocuments(registryText);
  } catch (error) {
    throw new RegistryLoadError(`${path} contains invalid YAML`, { cause: error });
  }

  if (documents.length !== 1) {
    throw new RegistryLoadError(`${path} must contain exactly one YAML document`);
  }

  const [document] = documents;
  if (document.errors.length > 0) {
    throw new RegistryLoadError(`${path} contains invalid YAML`, {
      cause: document.errors[0],
    });
  }

  let registryData: unknown;

  try {
    registryData = document.toJSON();
  } catch (error) {
    throw new RegistryLoadError(`${path} contains invalid YAML`, { cause: error });
  }

  return parseRegistryData(registryData);
}
